This patch does several things to make the Scotch build system work for Debian:
  * Default to building scotch and ptscotch
  * Clean up installation
  * Fix clean targets to remove shared libraries
  * Fix library and executable linking commands to work with binutils-gold and
    allow linking with both libscotch and libptscotch libraries

Forwarded: no
Reviewed-By: "Adam C. Powell, IV" <hazelsct@debian.org>
Last-Update: 2011-04-01

Index: scotch/src/Makefile
===================================================================
--- scotch.orig/src/Makefile	2021-06-27 23:11:26.531693643 +0200
+++ scotch/src/Makefile	2021-06-27 23:11:26.523693441 +0200
@@ -36,7 +36,7 @@
 
 .PHONY				:	clean	default	install	ptscotch	realclean	required	scotch
 
-default				:	scotch
+default				:	scotch ptscotch
 
 required			:	Makefile.inc	../bin	../include	../lib
 
@@ -97,14 +97,14 @@
 
 scotch				:	libscotch
 					(cd scotch ;         $(MAKE) VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL) scotch && $(MAKE) install)
-					(cd libscotchmetis ; $(MAKE)                                                                scotch && $(MAKE) install)
+					(cd libscotchmetis ; $(MAKE) VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL) scotch && $(MAKE) install)
 
 libptscotch			:	libscotch
 					(cd libscotch ;      $(MAKE) VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL) ptscotch && $(MAKE) ptinstall)
 
 ptscotch			:	libptscotch
 					(cd scotch ;         $(MAKE) VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL) ptscotch && $(MAKE) ptinstall)
-					(cd libscotchmetis ; $(MAKE)                                                                ptscotch && $(MAKE) ptinstall)
+					(cd libscotchmetis ; $(MAKE) VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL) ptscotch && $(MAKE) ptinstall)
 
 check				:	scotch
 					(cd check ; $(MAKE) check)
@@ -119,22 +119,19 @@
 					(cd check ; $(MAKE) ptcheck_full)
 
 esmumps				:	scotch
-					(cd esmumps ; $(MAKE) scotch && $(MAKE) install)
+					(cd esmumps ; $(MAKE) VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL) scotch && $(MAKE) VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL) install)
 
 escheck				:	esmumps
 					(cd check ; $(MAKE) escheck)
 
 ptesmumps			:	scotch	ptscotch
-					(cd esmumps ; $(MAKE) ptscotch && $(MAKE) ptinstall)
+					(cd esmumps ; $(MAKE) VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL) ptscotch && $(MAKE) VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL) ptinstall)
 
 install				:	required	$(bindir)	$(includedir)	$(libdir)	$(mandir)/man1
-					-$(CP) -f ../bin/[agm]*$(EXE) $(bindir)
-					-$(CP) -f ../bin/d[agm]*$(EXE) $(bindir)
-					-$(CP) -f ../include/*esmumps*.h $(includedir)
-					-$(CP) -f ../include/*scotch*.h $(includedir)
-					-$(CP) -f ../lib/*esmumps*$(LIB) $(libdir)
-					-$(CP) -f ../lib/*scotch*$(LIB) $(libdir)
-					-$(CP) -f ../man/man1/*[.]1 $(mandir)/man1
+					$(CP) -af ../bin/* $(bindir)
+					$(CP) -af ../include/*.h $(includedir)
+					$(CP) -af ../lib/lib* $(libdir)
+					$(CP) -af ../man/* $(mandir)
 
 installstub			:	required	$(includedir)
 					-$(CP) -f ../include/*metis*.h $(includestubdir)
Index: scotch/src/libscotchmetis/Makefile
===================================================================
--- scotch.orig/src/libscotchmetis/Makefile	2021-06-27 23:11:26.531693643 +0200
+++ scotch/src/libscotchmetis/Makefile	2021-06-27 23:12:09.056744382 +0200
@@ -67,15 +67,15 @@
 					libptscotchparmetis$(LIB)
 
 install				:	scotch
-					-$(CP) metis.h metisf.h $(includedir)
-					-$(CP) libscotchmetis$(LIB) $(libdir)
+					-$(CP) -a metis.h metisf.h $(includedir)
+					-$(CP) -a libscotchmetis$(LIB) libscotchmetis*.so $(libdir)
 
 ptinstall			:	ptscotch
-					-$(CP) parmetis.h $(includedir)
-					-$(CP) libptscotchparmetis$(LIB) $(libdir)
+					-$(CP) -a parmetis.h $(includedir)
+					-$(CP) -a libptscotchparmetis$(LIB) libptscotchparmetis*.so $(libdir)
 
 clean				:
-					-$(RM) *~ *$(OBJ) lib*$(LIB) metis.h metisf.h parmetis.h
+					-$(RM) *~ *$(OBJ) lib*$(LIB) lib*.so lib*.a metis.h metisf.h parmetis.h
 
 realclean			:	clean
 
@@ -187,6 +187,9 @@
 					parmetis_dgraph_order_f$(OBJ)		\
 					parmetis_dgraph_part$(OBJ)		\
 					parmetis_dgraph_part_f$(OBJ)
+					$(CC) -shared -Wl,-soname,libptscotchparmetis-$(VERSION).$(RELEASE).so $^ -L../libscotch -lptscotch -lptscotcherr -lscotch -lm -o libptscotchparmetis-$(VERSION).$(RELEASE).$(PATCHLEVEL).so
+					ln -s libptscotchparmetis-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libptscotchparmetis-$(VERSION).$(RELEASE).so
+					ln -s libptscotchparmetis-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libptscotchparmetis.so
 					$(AR) $(ARFLAGS) $(@) $(^)
 					-$(RANLIB) $(@)
 
@@ -200,5 +203,8 @@
 					metis_graph_part_dual_f$(OBJ)		\
 					metis_options$(OBJ)			\
 					metis_options_f$(OBJ)
+					$(CC) -shared -Wl,-soname,libscotchmetis-$(VERSION).$(RELEASE).so $^ -L../libscotch -lscotch -lscotcherr -lm -o libscotchmetis-$(VERSION).$(RELEASE).$(PATCHLEVEL).so
+					ln -s libscotchmetis-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libscotchmetis-$(VERSION).$(RELEASE).so
+					ln -s libscotchmetis-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libscotchmetis.so
 					$(AR) $(ARFLAGS) $(@) $(^)
 					-$(RANLIB) $(@)
Index: scotch/src/libscotch/Makefile
===================================================================
--- scotch.orig/src/libscotch/Makefile	2021-06-27 23:11:26.531693643 +0200
+++ scotch/src/libscotch/Makefile	2021-06-27 23:11:26.527693541 +0200
@@ -89,7 +89,7 @@
 					$(libdir)/libptscotch$(LIB)
 
 clean				:
-					-$(RM) *~ *$(OBJ) lib*$(LIB) parser_yy.c parser_ly.h parser_ll.c *scotch.h *scotchf.h y.output *dummysizes$(EXE)
+					-$(RM) *~ *$(OBJ) lib*.a lib*.so* parser_yy.c parser_ly.h parser_ll.c *scotch.h *scotchf.h y.output *dummysizes$(EXE)
 
 realclean			:	clean
 
@@ -99,22 +99,22 @@
 ##
 
 $(includedir)/scotch.h		:	scotch.h
-					-$(CP) scotch.h $(includedir)
+					$(CP) -a scotch.h $(includedir)
 
 $(includedir)/scotchf.h		:	scotchf.h
-					-$(CP) scotchf.h $(includedir)
+					$(CP) -a scotchf.h $(includedir)
 
 $(includedir)/ptscotch.h	:	ptscotch.h
-					-$(CP) ptscotch.h $(includedir)
+					$(CP) -a ptscotch.h $(includedir)
 
 $(includedir)/ptscotchf.h	:	ptscotchf.h
-					-$(CP) ptscotchf.h $(includedir)
+					$(CP) -a ptscotchf.h $(includedir)
 
 $(libdir)/libscotch$(LIB)	:	libscotch$(LIB)
-					-$(CP) libscotch$(LIB) libscotcherr*$(LIB) $(libdir)
+					$(CP) -a libscotch* $(libdir)
 
 $(libdir)/libptscotch$(LIB)	:	libptscotch$(LIB)
-					-$(CP) libptscotch$(LIB) libptscotcherr*$(LIB) $(libdir)
+					$(CP) -a libptscotch* $(libdir)
 
 ##
 ##  Library dependencies.
@@ -3121,26 +3121,44 @@
 					library_pt_f.h
 					./ptdummysizes$(EXE) "-s$(SCOTCH_NAME_SUFFIX)" library_pt_f.h ptscotchf.h
 
-libscotch$(LIB)			:	$(LIBSCOTCHDEPS)
-					$(AR) $(ARFLAGS) $(@) $(?)
+libscotch$(LIB)			:	$(LIBSCOTCHDEPS) libscotcherr$(LIB)
+					$(CC) -shared -Wl,-soname,libscotch-$(VERSION).$(RELEASE).so $(LIBSCOTCHDEPS) -L. -lscotcherr -lz -lbz2 -llzma -lpthread -lm -lrt -o libscotch-$(VERSION).$(RELEASE).$(PATCHLEVEL).so
+					ln -s libscotch-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libscotch-$(VERSION).$(RELEASE).so
+					ln -s libscotch-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libscotch.so
+					$(AR) $(ARFLAGS) $(@) $(LIBSCOTCHDEPS)
 					-$(RANLIB) $(@)
 
 libscotcherr$(LIB)		:	library_error$(OBJ)
+					$(CC) -shared -Wl,-soname,libscotcherr-$(VERSION).$(RELEASE).so $^ -o libscotcherr-$(VERSION).$(RELEASE).$(PATCHLEVEL).so
+					ln -s libscotcherr-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libscotcherr-$(VERSION).$(RELEASE).so
+					ln -s libscotcherr-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libscotcherr.so
 					$(AR) $(ARFLAGS) $(@) $(?)
 					-$(RANLIB) $(@)
 
 libscotcherrexit$(LIB)		:	library_error_exit$(OBJ)
+					$(CC) -shared -Wl,-soname,libscotcherrexit-$(VERSION).$(RELEASE).so $^ -o libscotcherrexit-$(VERSION).$(RELEASE).$(PATCHLEVEL).so
+					ln -s libscotcherrexit-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libscotcherrexit-$(VERSION).$(RELEASE).so
+					ln -s libscotcherrexit-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libscotcherrexit.so
 					$(AR) $(ARFLAGS) $(@) $(?)
 					-$(RANLIB) $(@)
 
-libptscotch$(LIB)		:	$(LIBPTSCOTCHDEPS)
-					$(AR) $(ARFLAGS) $(@) $(?)
+libptscotch$(LIB)		:	$(LIBPTSCOTCHDEPS) $(LIBSCOTCHDEPS) libptscotcherr$(LIB)
+					$(CC) -shared -Wl,-soname,libptscotch-$(VERSION).$(RELEASE).so $(LIBPTSCOTCHDEPS) $(LIBSCOTCHDEPS) -L. -lptscotcherr -lz -lbz2 -llzma -lpthread -lm -lrt -o libptscotch-$(VERSION).$(RELEASE).$(PATCHLEVEL).so
+					ln -s libptscotch-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libptscotch-$(VERSION).$(RELEASE).so
+					ln -s libptscotch-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libptscotch.so
+					$(AR) $(ARFLAGS) $(@) $(LIBPTSCOTCHDEPS) $(LIBSCOTCHDEPS)
 					-$(RANLIB) $(@)
 
 libptscotcherr$(LIB)		:	library_error_pt$(OBJ)
+					$(CC) -shared -Wl,-soname,libptscotcherr-$(VERSION).$(RELEASE).so $^ -o libptscotcherr-$(VERSION).$(RELEASE).$(PATCHLEVEL).so
+					ln -s libptscotcherr-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libptscotcherr-$(VERSION).$(RELEASE).so
+					ln -s libptscotcherr-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libptscotcherr.so
 					$(AR) $(ARFLAGS) $(@) $(?)
 					-$(RANLIB) $(@)
 
 libptscotcherrexit$(LIB)	:	library_error_exit_pt$(OBJ)
+					$(CC) -shared -Wl,-soname,libptscotcherrexit-$(VERSION).$(RELEASE).so $^ -o libptscotcherrexit-$(VERSION).$(RELEASE).$(PATCHLEVEL).so
+					ln -s libptscotcherrexit-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libptscotcherrexit-$(VERSION).$(RELEASE).so
+					ln -s libptscotcherrexit-$(VERSION).$(RELEASE).$(PATCHLEVEL).so libptscotcherrexit.so
 					$(AR) $(ARFLAGS) $(@) $(?)
 					-$(RANLIB) $(@)
Index: scotch/src/esmumps/Makefile
===================================================================
--- scotch.orig/src/esmumps/Makefile	2021-06-27 23:11:26.531693643 +0200
+++ scotch/src/esmumps/Makefile	2021-06-27 23:11:26.527693541 +0200
@@ -45,7 +45,8 @@
 				$(CC) $(CFLAGS) $(CLIBFLAGS) -I$(srclibscotchdir) -I$(includedir) -c $(<) -o $(@)
 
 %$(EXE)	:	%.c
-		 		$(CC) $(CFLAGS) -I$(srclibscotchdir) -I$(includedir) $(<) -o $(@) -L. -lesmumps -L$(libdir) -lscotch -lscotcherrexit $(LDFLAGS)
+		 		$(CC) $(CFLAGS) -I$(srclibscotchdir) -I$(includedir) $(<) -o $(@) -L. -l$(ESMUMPSLIB) -L$(libdir) -l$(SCOTCHLIB) -l$(SCOTCHLIB)errexit $(LDFLAGS)
+
 
 ##
 ##  Project rules.
@@ -54,24 +55,32 @@
 .PHONY				:	ptscotch	scotch	ptinstall	install	clean	realclean
 
 scotch				:	clean
-					$(MAKE) CC="$(CCS)" CCD="$(CCS)"	\
+					$(MAKE) CC="$(CCS)" CCD="$(CCS)" SCOTCHLIB=scotch ESMUMPSLIB=esmumps	\
+					VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL)		\
 					libesmumps$(LIB)			\
 					main_esmumps$(EXE)
 
-ptscotch			:	scotch
+ptscotch			:	clean
+					$(MAKE) CFLAGS="$(CFLAGS) -DSCOTCH_PTSCOTCH" CC="$(CCP)" SCOTCHLIB=ptscotch ESMUMPSLIB=ptesmumps	\
+					VERSION=$(VERSION) RELEASE=$(RELEASE) PATCHLEVEL=$(PATCHLEVEL)						\
+					libptesmumps$(LIB)											\
+					main_ptesmumps$(EXE)
 
 install				:	scotch					\
 					library.h
 					-$(CP) library.h $(includedir)/esmumps.h
-					-$(CP) libesmumps$(LIB) $(libdir)
+					-$(CP) -a libesmumps* $(libdir)
+					-$(CP) -a main_esmumps$(EXE) $(bindir)/scotch_esmumps
 
-ptinstall			:	scotch					\
+ptinstall			:	ptscotch					\
 					library.h
-					-$(CP) library.h $(includedir)/esmumps.h
-					-$(CP) libesmumps$(LIB) $(libdir)/libptesmumps$(LIB)
+					-$(CP) library.h $(includedir)/ptesmumps.h
+					-$(CP) -a libptesmumps* $(libdir)
+					-$(CP) -a main_ptesmumps$(EXE) $(bindir)/ptscotch_esmumps
+
 
 clean				:
-					-$(RM) *~ common.h *$(OBJ) lib*$(LIB) main_esmumps$(EXE)
+					-$(RM) *~ common.h *$(OBJ) lib*$(LIB) lib*.so main_*esmumps$(EXE)
 
 realclean			:	clean
 
@@ -195,7 +204,10 @@
 					order.h					\
 					fax.h
 
-libesmumps$(LIB)		:	$(LIBESMUMPSDEPS)
+lib$(ESMUMPSLIB)$(LIB)		:	$(LIBESMUMPSDEPS)
+					$(CC) -shared -Wl,-soname,lib$(ESMUMPSLIB)-$(VERSION).$(RELEASE).so $^ -L../libscotch -l$(SCOTCHLIB) -l$(SCOTCHLIB)err -o lib$(ESMUMPSLIB)-$(VERSION).$(RELEASE).$(PATCHLEVEL).so
+					ln -s lib$(ESMUMPSLIB)-$(VERSION).$(RELEASE).$(PATCHLEVEL).so lib$(ESMUMPSLIB)-$(VERSION).$(RELEASE).so
+					ln -s lib$(ESMUMPSLIB)-$(VERSION).$(RELEASE).$(PATCHLEVEL).so lib$(ESMUMPSLIB).so
 					$(AR) $(ARFLAGS) $(@) $(?)
 					-$(RANLIB) $(@)
 
@@ -206,6 +218,12 @@
 					order.h					\
 					symbol.h				\
 					esmumps.h				\
-					libesmumps$(LIB)			\
-					$(libdir)/libscotch$(LIB)		\
-					$(libdir)/libscotcherrexit$(LIB)
+					lib$(ESMUMPSLIB)$(LIB)			\
+					$(libdir)/lib$(SCOTCHLIB)$(LIB)		\
+					$(libdir)/lib$(SCOTCHLIB)errexit$(LIB)
+
+main_ptesmumps$(EXE)		:
+					-$(MV) main_esmumps$(EXE) main_tmp_esmumps$(EXE) 2>/dev/null || /bin/true
+					$(MAKE) CFLAGS="$(CFLAGS) -DSCOTCH_PTSCOTCH" CC="$(CCP)" SCOTCHLIB=ptscotch ESMUMPSLIB=ptesmumps main_esmumps$(EXE)
+					$(MV) -f main_esmumps$(EXE) main_ptesmumps$(EXE)
+					-$(MV) main_tmp_esmumps$(EXE) main_esmumps$(EXE) 2>/dev/null || /bin/true
Index: scotch/src/scotch/Makefile
===================================================================
--- scotch.orig/src/scotch/Makefile	2021-06-27 23:11:26.531693643 +0200
+++ scotch/src/scotch/Makefile	2021-06-27 23:11:26.527693541 +0200
@@ -41,13 +41,13 @@
 include ../Makefile.inc
 
 %$(EXE)	:	%$(OBJ)
-		 		$(CC) $(CFLAGS) -I$(includedir) -I../libscotch $(<) -o $(@) -L$(libdir) -l$(SCOTCHLIB) -lscotch -l$(SCOTCHLIB)errexit $(LDFLAGS)
+		 		$(CC) $(CFLAGS) -I$(includedir) -I../libscotch $(<) -o $(@) -L$(libdir) -l$(SCOTCHLIB) -lscotcherr -l$(SCOTCHLIB)errexit $(LDFLAGS)
 
 %$(OBJ)	:	%.c
 				$(CC) $(CFLAGS) -I$(includedir) -I../libscotch -c $(<) -o $(@)
 
 %$(EXE)	:	%.c
-		 		$(CC) $(CFLAGS) -I$(includedir) -I../libscotch $(<) -o $(@) -L$(libdir) -l$(SCOTCHLIB) -lscotch -l$(SCOTCHLIB)errexit $(LDFLAGS)
+		 		$(CC) $(CFLAGS) -I$(includedir) -I../libscotch $(<) -o $(@) -L$(libdir) -l$(SCOTCHLIB) -lscotcherr -l$(SCOTCHLIB)errexit $(LDFLAGS)
 
 ##
 ##  Project rules.
@@ -330,7 +330,7 @@
 					gout_o.c				\
 					$(includedir)/scotch.h			\
 					$(libdir)/libscotch$(LIB)
-					$(CC) $(CFLAGS) -I../libscotch -I$(includedir) gout_c.c gout_o.c -o $(@) -L$(libdir) -lscotch -lscotcherrexit $(LDFLAGS)
+					$(CC) $(CFLAGS) -I../libscotch -I$(includedir) gout_c.c gout_o.c -o $(@) -L$(libdir) -l$(SCOTCHLIB) -lscotcherr -lscotcherrexit $(LDFLAGS)
 
 gpart$(EXE)			:	gmap.c					\
 					../libscotch/module.h			\
@@ -339,7 +339,7 @@
 					$(libdir)/libscotch$(LIB)		\
 					$(libdir)/libscotcherrexit$(LIB)	\
 					gmap.h
-			 		$(CC) $(CFLAGS) -I$(includedir) -I../libscotch $(<) -DSCOTCH_COMPILE_PART -o $(@) -L$(libdir) -lscotch -lscotcherrexit $(LDFLAGS)
+			 		$(CC) $(CFLAGS) -I$(includedir) -I../libscotch $(<) -DSCOTCH_COMPILE_PART -o $(@) -L$(libdir) -l$(SCOTCHLIB) -lscotcherr -lscotcherrexit $(LDFLAGS)
 
 gscat$(EXE)			:	gscat.c					\
 					../libscotch/module.h			\
